### ISSUE TITLE
Proposal for shortcut operator with ternary evaluation

### ISSUE THREAD

**Issue by jleeothon** [cortexlang/cortex#139]
As an idiom for "set value if it's null", we considered the following syntax inspired in ternary operators and shortcut assignment operators:

This proposal includes optional information relating to #6.

```
a ?= b
// equivalent to a = a ? a : b
// or if compatible with the above syntax, a = a ? ? : b ***
// looks funny
```

As an extension, we may want "initialize to the first non-null (or non false) value of n values".
```
a ?= b : c
// a = a ? a : (b ? b : c)
// a = a ? ? : (b ? ? : c) ***
// I wonder if parser would handle a = a?? : (b?? c) out of the box

a ?= b : c : d : e
a = a ? ? : (b ? ? : (c ? ? : (d ? ? : e))) ***
```

I wonder what the context variable `?` would mean in the above cases.
In the case of the expanded versions marked `***`, the ternary operator is a shortcut for the innermost condition variable.
In the case of the shortcut operator `?=`, however, it sounds as though `?` should *always* be replaced by the left-hand side of the assignment.

**Comment by SanderMertens** [comment]
The above examples mix two different ideas (left hand side evaluation and right hand side evaluation). These may be unified in a single syntax, or not. However, the above cannot be implemented due to ambiguity:

Does `a ?= b : c` mean `if a: a = b else a = c` or `if b: a = b else a = c`.

**Comment by jleeothon** [comment]
Which ambiguity? The example calls specifically for the first option, or am I missing something?

**Comment by SanderMertens** [comment]
Then, what does the statement below mean (from the comment)?
`a ?= b : c : d : e`

**Comment by jleeothon** [comment]
From the description, it would mean
```
a = a ? (b ? b : (c ? c : (d ? d : e)))
```

**Comment by jleeothon** [comment]
We could identify two issues with the above approach:
- The operator resembles the shortcut binary operators, but this one is not binary, so it doesn't work as straightforward as other operators.
- It forces you to take the left-hand side of the assignment and you could not shortcut without it. If we specify that the left-hand side is not part of the false-checking, it now looks confusing.

However, drawing influence from the C# `??` operator, we could make a less ambiguous operator. Take:
```
a = a ?? b // a is a or b // it's less dry in this use-case but it's not overly verbose either
a = b ?? c // a is b or c // a is not forced into consideration
```

In C#, this operator can be chained, which is a good idea to me.

```
a = b ?? c ?? d // a is the first non-null/non-false of b, c or d
a = a ?? b ?? c ?? d // same as before, but doesn't change if a already has a value
```

The only point that I would like to extend is the operator types. Apparently, it only works as a `null` check. Maybe we can use it as a `false` check so it also works with primitives (it doesn't work with C# (non-nullable) structs).

In short, it could work similar to Python's `or` operator.

**Comment by SanderMertens** [comment]
To summarize different kinds of 'conditional' operators, as encountered in different languages:
```cond ? b : c // return b if cond == true, else c```
`a ?= b : c // assign b if a == true, else assign c`
`a = b ?? c // assign b if b, else assign c`
`a ??= b // assign b if a`
`a?.b?.c?.d // dereference b if a, c if b, d if c`

I'd like to come up with a convention for all of the above. For example, it is very obvious that the `+=` operator means "add and assign" because of a simple convention. A similar convention is desirable here since the `??` operator can look as confusing as its name suggests.

**Comment by jleeothon** [comment]
In your list,
- The first exists in C, Java, etc.
- The second one doesn't exist, does it?
- The third one exists in C#.
- The fourth one doesn't exist, does it?
- The fifth one exists in C#.

**Comment by SanderMertens** [comment]
Correct. The above list extrapolates existing operators to assignment-counterparts. Thoughts on how to make the above look more consistent?
